### PR TITLE
RTEMIS-187: Added exposed filter.

### PR DESCRIPTION
**Pull request by RohanSingh-inno**
Ticket # : https://innoraft.atlassian.net/browse/RTEMIS-187

---
### What the PR does
`(Explain fixes, changes and impacts)`
-  Added exposed filters.
-  Added academic cycle constraint.
-  Added a routeSubscriber for allocation edit check.

### CODE REVIEW

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+   * Checks access to the user register page based on academic_session.
+   */
+  public function access(AccountInterface $account, RouteMatchInterface $routeMatch) {
+    $miniNode = $routeMatch->getParameter('mini_node') ?? NULL;
+    if ($miniNode instanceof EckEntityInterface && $miniNode->bundle() == 'allocation' && $account->hasPermission('edit any mini_node entities of bundle allocation')) {
+      $uid = $account->id();
+      $userEntity = $this->entityTypeManager->getStorage('user')->load($uid);
+      if ($userEntity instanceof UserInterface) {
+        $schoolId = $userEntity->get('field_school_details')->getString() ?? NULL;
+        if ($schoolId != $miniNode) {
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php:56]: Here `$miniNode` is entity and not id right? then why are we comparing it with `schoolId`.

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+   * Constructs an UserRegisterAccessCheck object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   Config factory.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager, ConfigFactoryInterface $config_factory) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->configFactory = $config_factory;
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php:43]: if we are not using `$this->configFactory`, then please remove this.

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Routing\Access\AccessInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\eck\EckEntityInterface;
+use Drupal\user\UserInterface;
+
+/**
+ * Determines edit access to the school details mini node.
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php:15]: Description doesn't match the functionality.

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   Config factory.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager, ConfigFactoryInterface $config_factory) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * Checks access to the user register page based on academic_session.
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php:47]: Description does match the functionality here as well

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+   */
+  public function access(AccountInterface $account, RouteMatchInterface $routeMatch) {
+    $miniNode = $routeMatch->getParameter('mini_node') ?? NULL;
+    if ($miniNode instanceof EckEntityInterface && $miniNode->bundle() == 'allocation' && $account->hasPermission('edit any mini_node entities of bundle allocation')) {
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php:51]: Please provide the `edit any mini_node entities of bundle allocation` permission to school_admin.

Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php:40]: ok

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php (hunk 4 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php]: modified

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php (hunk 3 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php]: modified

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php (hunk 2 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php]: ya sure we can remove this.

File: modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php (hunk 1 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Access/AllocationDetailEditAccessCheck.php]: modified

File: modules/rte_mis_allocation/rte_mis_allocation.module [hunk 6]
```diff
@@ -21,7 +21,7 @@ function rte_mis_allocation_get_allowed_values($definition, $entity = NULL) {
   $options = [];
   switch ($field_name) {
     case 'field_academic_year_allocation':
-      $options[] = _rte_mis_core_get_current_academic_year();
+      $options[_rte_mis_core_get_current_academic_year()] = str_replace('_', '-', _rte_mis_core_get_current_academic_year());
```
[the remark this answers is not in the document]
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/rte_mis_allocation.module]: yes we can.

**Pull request merged**